core/schema/container.go

```diff
@@ -1479,7 +1480,17 @@ func expandEnvVar(ctx context.Context, parent *core.Container, input string, exp
 		return "", err
 	}
 
+	secretEnvs := []string{}
+	for _, secret := range parent.Secrets {
+		secretEnvs = append(secretEnvs, secret.EnvName)
+	}
+
 	return os.Expand(input, func(k string) string {
+		// do not expand if its a secret env variable
+		if slices.Contains(secretEnvs, k) {
+			return "${" + k + "}"
+		}
```

Comment:
Huh this is a kinda neat idea - if we can't expand a var, leave it there.

That said, I think it's still a bit confusing - the case for `expand` is that there is no shell, and you want to expand it without one. So leaving shell-related expansion around in the result isn't great.

Maybe we should expand secrets as `***`? Censor them here if we really want this to succeed. Although, I still think that raising an error here makes more sense, since we can't really safely/generically expand secrets here at all.

Reply:
Hi @jedevc,

one thing I want to make sure is that you have read through this comment where I have tried to summarize the issue: https://github.com/dagger/dagger/issues/8752#issuecomment-2429864720

tl;dr; it seems the secrets expansion happens [here](https://github.com/dagger/dagger/blob/90befd2748d41b6c5112f9e296a3579a773471c9/core/container_exec.go#L212C8-L212C17), which happens after the expand function has been called already. 

so if we leave the environment variables which are suppose to be secrets-env-variables, then they will expand in the downstream function as expected.

does that make sense?

Reply:
> tl;dr; it seems the secrets expansion happens [here](https://github.com/dagger/dagger/blob/90befd2748d41b6c5112f9e296a3579a773471c9/core/container_exec.go#L212C8-L212C17), which happens after the expand function has been called already.

Secrets expansion is not actually done here. The code you link to only specifies to buildkit that those secrets exist, and provides them as environment variables - there is no expansion in dagger or buildkit. The reason that secrets are expanded at all is because the shell `sh` is doing it.

> so if we leave the environment variables which are suppose to be secrets-env-variables, then they will expand in the downstream function as expected.

Right, but only if there is a shell - if there is no shell, for instance you're doing `WithFile("${secret_name}")`, then this would just create a file called `${secret_name}` - which is confusing.

I think the better behavior is to just provide a neat error message to users if they expand a secret variable: skipping it and leaving it empty is confusing, expanding to `***` is possible, but also confusing/

Reply:
ah I see what you are saying. 

one of the intermediate revision I had was to implement a new function just for WithExec, where I keep the secret string as is, and for all other scenarios, we can error out if the secret variable is used in expand env.

I think that should cover both exec and other scenario 

OR

we can disable expandEnv in WithExec and rely on shell expansion.

Reply:
👍 , so I have now made change to not allow expanding Secret Variables when using Expand. It will now error out with a msg `expand cannot be used with secret env variable "FOO"`